File: Scripts/group.py
class Group:
    def __init__(self, name, leader=None):
        self.name = name
        self.members = []
        self.leader = leader
        self.relations = {}

    def add_member(self, agent):
        self.members.append(agent)

    def remove_member(self, agent):
        self.members.remove(agent)
        if(agent == self.leader):
            self.elect_leader()

    def group_action(self):
        # Define group-level actions. The action might depend on the leader's decision,
        # a consensus mechanism, or other factors.
        pass

    def elect_leader(self, candidates=None):
        if candidates is None:
            candidates = self.members
        
        votes = {}
        for member in self.members:
            vote = member.vote(candidates)
            print(member.name + " voted for " + vote.name)  # Assume each agent has a 'vote' method
            votes[vote] = votes.get(vote, 0) + 1
        
        max_votes = max(votes.values(), default=0)
        top_candidates = [agent for agent, vote_count in votes.items() if vote_count == max_votes]
        print(top_candidates)
        # If there's a single leader, or the current leader is among the tied candidates
        if len(top_candidates) == 1:
            self.leader = top_candidates[0]
            print(self.leader.name + " was elected as leader of " + self.name)
        elif (self.leader in top_candidates):
            print(self.leader.name + " was re-elected.")
        else:
            # Recurse with the tied members
            self.elect_leader(candidates=top_candidates)
            

File: src/types/group.ts
import { Agent } from "./agent"; // Import your Agent class

export class Group {
  name: string;
  members: Agent[] = [];
  leader: Agent | null = null;
  relations: Record<string, any> = {};

  constructor(name: string, leader: Agent | null = null) {
    this.name = name;
    this.leader = leader;
  }

  addMember(agent: Agent): void {
    this.members.push(agent);
  }

  removeMember(agent: Agent): void {
    this.members = this.members.filter(member => member !== agent);
    if (agent === this.leader) {
      this.electLeader();
    }
  }

  groupAction(): void {
    // Define group-level actions here
  }

  electLeader(candidates: Agent[] | null = null): void {
    if (!candidates) {
      candidates = this.members;
    }

    const votes: Map<Agent, number> = new Map();

    for (const member of this.members) {
      const vote = member.vote(candidates);
      console.log(`${member.name} voted for ${vote.name}`);
      votes.set(vote, (votes.get(vote) || 0) + 1);
    }

    const maxVotes = Math.max(...Array.from(votes.values()));
    const topCandidates = Array.from(votes.entries())
      .filter(([_, voteCount]) => voteCount === maxVotes)
      .map(([agent]) => agent);

    if (topCandidates.length === 1) {
      this.leader = topCandidates[0];
      console.log(`${this.leader.name} was elected as leader of ${this.name}`);
    } else if (this.leader && topCandidates.includes(this.leader)) {
      console.log(`${this.leader.name} was re-elected.`);
    } else {
      this.electLeader(topCandidates);
    }
  }
}
